src/DynamoCore/PackageManager/PackageUploadBuilder.cs

```diff
@@ -1,205 +1,93 @@
+        private readonly IFileCompressor fileCompressor;
 
-            return new PackageUploadRequestBody(l.Name, l.VersionName, l.Description, l.Keywords, l.License, l.Contents, "dynamo",
-                                                         engineVersion, engineMetadata, l.Group, l.Dependencies, 
-                                                         l.SiteUrl, l.RepositoryUrl, l.ContainsBinaries, l.NodeLibraries.Select(x => x.FullName) ); 
-        } 
+        internal const string PackageEngineName = "dynamo";
+        internal const long MaximumPackageSize = 100 * 1024 * 1024;
 
-        public static PackageUpload NewPackage(string rootPkgDir, CustomNodeManager customNodeManager, Package pkg, List<string> files, PackageUploadHandle uploadHandle, bool isTestMode)
+        internal PackageUploadBuilder(PackageDirectoryBuilder builder, IFileCompressor fileCompressor)
         {
-            var header = NewPackageHeader(pkg);
-            var zipPath = DoPackageFileOperationsAndZip(rootPkgDir, customNodeManager, header, pkg, files, uploadHandle, isTestMode);
-            return new PackageUpload(header, zipPath);
+            if (builder == null) throw new ArgumentNullException("builder");
+            if (fileCompressor == null) throw new ArgumentNullException("fileCompressor");
+
+            this.builder = builder;
+            this.fileCompressor = fileCompressor;
         }
+         
+        #region Public Operational Class Methods
 
-        public static PackageVersionUpload NewPackageVersion(string rootPkgDir, CustomNodeManager customNodeManager, Package pkg, List<string> files, PackageUploadHandle uploadHandle, bool isTestMode)
+        internal static PackageUploadRequestBody NewRequestBody(Package l)
         {
-            var header = NewPackageHeader(pkg);
-            var zipPath = DoPackageFileOperationsAndZip(rootPkgDir, customNodeManager, header, pkg, files, uploadHandle, isTestMode);
-            return new PackageVersionUpload(header, zipPath);
+            var engineVersion = Assembly.GetExecutingAssembly().GetName().Version.ToString();
+            var engineMetadata = "";
+
+            return new PackageUploadRequestBody(l.Name, l.VersionName, l.Description, l.Keywords, l.License, l.Contents, PackageEngineName,
+                                                         engineVersion, engineMetadata, l.Group, l.Dependencies,
+                                                         l.SiteUrl, l.RepositoryUrl, l.ContainsBinaries, l.NodeLibraries.Select(x => x.FullName));
         }
 
-    #region Utility methods
 
-        private static string DoPackageFileOperationsAndZip(string rootPkgDir, CustomNodeManager customNodeManager, PackageUploadRequestBody header, Package pkg, List<string> files, PackageUploadHandle uploadHandle, bool isTestMode)
+        internal PackageUpload NewPackageUpload(Package package, string packagesDirectory, IEnumerable<string> files, PackageUploadHandle handle)
```

Comment:
`DoPackageFileOperationsAndZip` is moved to `PackageDirectoryBuilder` as `BuildDirectory`.  File compression, which is necessary for the upload, is moved here.